File: js/packages/app/src/__tests__/features/object-references/external/wrappers/object-references-example/src/polywrap/internal/global-functions/invokeGlobalFunction.ts
import { parse } from '@serial-as/json'
import { bufferToU32, BaseTypeSerialization, IExternalWrapInstance } from "@nerfzael/reim-wrap-as";
import { WrapManifest } from '../../WrapManifest';
import { 
  
  testReceiveReference,
    
  testInvokeExternalGlobalFunction,
    
  testInvokeExternalStaticMethod,
    
  testInvokeExternalInstanceMethod,
       
} from "../../..";
import { TestExternalClassWrapped } from "../../wrapped";
import { TestExternalClass } from "../../external";



export function invoke(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer {
  const funcId = bufferToU32(buffer);
  const dataBuffer = buffer.slice(4);

  switch (funcId) {
    
    case WrapManifest.Internal.GlobalFunction.TestReceiveReference:
        return invokeTestReceiveReferenceWrapped(dataBuffer, wrapInstance);
        
    case WrapManifest.Internal.GlobalFunction.TestInvokeExternalGlobalFunction:
        return invokeTestInvokeExternalGlobalFunctionWrapped(dataBuffer, wrapInstance);
        
    case WrapManifest.Internal.GlobalFunction.TestInvokeExternalStaticMethod:
        return invokeTestInvokeExternalStaticMethodWrapped(dataBuffer, wrapInstance);
        
    case WrapManifest.Internal.GlobalFunction.TestInvokeExternalInstanceMethod:
        return invokeTestInvokeExternalInstanceMethodWrapped(dataBuffer, wrapInstance);
            
    default:
      throw new Error("Unknown function/method: " + funcId.toString());
  }
}

const invokeTestReceiveReferenceWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const argsBuffer = buffer;

  const args = TestReceiveReferenceArgsWrapped.deserialize(argsBuffer, wrapInstance);

  const result = testReceiveReference(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestReceiveReferenceArgs {
  constructor(
    public arg: TestExternalClass,
  ) {
  }
}

@serializable
class TestReceiveReferenceArgsWrapped {
  constructor(
    public arg: TestExternalClassWrapped,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestReceiveReferenceArgs {
    const args = parse<TestReceiveReferenceArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestReceiveReferenceArgs(
      
      TestExternalClassWrapped.mapFromSerializable(args.arg, wrapInstance),
            
    );
  }  
}
const invokeTestInvokeExternalGlobalFunctionWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const argsBuffer = buffer;

  const args = TestInvokeExternalGlobalFunctionArgsWrapped.deserialize(argsBuffer, wrapInstance);

  const result = testInvokeExternalGlobalFunction(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestInvokeExternalGlobalFunctionArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class TestInvokeExternalGlobalFunctionArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestInvokeExternalGlobalFunctionArgs {
    const args = parse<TestInvokeExternalGlobalFunctionArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestInvokeExternalGlobalFunctionArgs(
            
      args.arg,
      
    );
  }  
}
const invokeTestInvokeExternalStaticMethodWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const argsBuffer = buffer;

  const args = TestInvokeExternalStaticMethodArgsWrapped.deserialize(argsBuffer, wrapInstance);

  const result = testInvokeExternalStaticMethod(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestInvokeExternalStaticMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class TestInvokeExternalStaticMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestInvokeExternalStaticMethodArgs {
    const args = parse<TestInvokeExternalStaticMethodArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestInvokeExternalStaticMethodArgs(
            
      args.arg,
      
    );
  }  
}
const invokeTestInvokeExternalInstanceMethodWrapped = (buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): ArrayBuffer => {
  const argsBuffer = buffer;

  const args = TestInvokeExternalInstanceMethodArgsWrapped.deserialize(argsBuffer, wrapInstance);

  const result = testInvokeExternalInstanceMethod(
    args.arg,
  );

  
  return BaseTypeSerialization.serialize<string>(result);
};

class TestInvokeExternalInstanceMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

@serializable
class TestInvokeExternalInstanceMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static deserialize(buffer: ArrayBuffer, wrapInstance: IExternalWrapInstance): TestInvokeExternalInstanceMethodArgs {
    const args = parse<TestInvokeExternalInstanceMethodArgsWrapped>(String.UTF8.decode(buffer));
  
    return new TestInvokeExternalInstanceMethodArgs(
            
      args.arg,
      
    );
  }  
}

File: rust/app/src/wrappers/external_references/src/polywrap/internal/invoke_global_function.rs
use std::sync::Arc;
use std::str;

use reim_dt::ExternalModule;
use serde::{Deserialize, Serialize};
use crate::polywrap::wrap::wrap_manifest;
use crate::{ 
  
  testReceiveReference,
    
  testInvokeExternalGlobalFunction,
    
  testInvokeExternalStaticMethod,
    
  testInvokeExternalInstanceMethod,
       
};
use crate::polywrap::external::wrapped::TestExternalClassWrapped;
use crate::polywrap::internal::wrapped::StringWrapped;
use crate::polywrap::external::TestExternalClass;


pub fn invoke(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
  let func_id = u32::from_be_bytes(buffer.try_into().expect("Function ID must be 4 bytes"));
  let data_buffer = &buffer[4..];

  match func_id {
    
    x if x == wrap_manifest::internal::GlobalFunction::TestReceiveReference as u32 =>
        invokeTestReceiveReferenceWrapped(data_buffer, external_module),
        
    x if x == wrap_manifest::internal::GlobalFunction::TestInvokeExternalGlobalFunction as u32 =>
        invokeTestInvokeExternalGlobalFunctionWrapped(data_buffer, external_module),
        
    x if x == wrap_manifest::internal::GlobalFunction::TestInvokeExternalStaticMethod as u32 =>
        invokeTestInvokeExternalStaticMethodWrapped(data_buffer, external_module),
        
    x if x == wrap_manifest::internal::GlobalFunction::TestInvokeExternalInstanceMethod as u32 =>
        invokeTestInvokeExternalInstanceMethodWrapped(data_buffer, external_module),
            
    _ => panic!("Unknown internal global function ID: {}", func_id.to_string()),
  }
}

fn invokeTestReceiveReferenceWrapped(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
  let args_buffer = buffer;

  let args = TestReceiveReferenceArgsWrapped::deserialize(args_buffer, external_module);

  let result = testReceiveReference(
    args.arg,
  );

  
  StringWrapped::serialize(&result).to_vec()
}

struct TestReceiveReferenceArgs {
    pub arg: Arc<TestExternalClass>,
}
 
impl TestReceiveReferenceArgs {
    pub fn new(
        arg: Arc<TestExternalClass>,
    ) -> Self {
        Self {
            arg,
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct TestReceiveReferenceArgsWrapped {
    pub arg: TestExternalClassWrapped,
}

impl TestReceiveReferenceArgsWrapped {
    pub fn new(
        arg: TestExternalClassWrapped,
    ) -> Self {
        Self {
            arg,
        }
    }

    pub fn deserialize(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> TestReceiveReferenceArgs {
        let args: TestReceiveReferenceArgsWrapped = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
       
        TestReceiveReferenceArgs::new(
            
            TestExternalClassWrapped::map_from_serializable(&args.arg, external_module),
                        
        )
    }  
}
fn invokeTestInvokeExternalGlobalFunctionWrapped(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
  let args_buffer = buffer;

  let args = TestInvokeExternalGlobalFunctionArgsWrapped::deserialize(args_buffer, external_module);

  let result = testInvokeExternalGlobalFunction(
    args.arg,
  );

  
  StringWrapped::serialize(&result).to_vec()
}

struct TestInvokeExternalGlobalFunctionArgs {
    pub arg: String,
}
 
impl TestInvokeExternalGlobalFunctionArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct TestInvokeExternalGlobalFunctionArgsWrapped {
    pub arg: String,
}

impl TestInvokeExternalGlobalFunctionArgsWrapped {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }

    pub fn deserialize(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> TestInvokeExternalGlobalFunctionArgs {
        let args: TestInvokeExternalGlobalFunctionArgsWrapped = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
       
        TestInvokeExternalGlobalFunctionArgs::new(
                        
            args.arg,
            
        )
    }  
}
fn invokeTestInvokeExternalStaticMethodWrapped(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
  let args_buffer = buffer;

  let args = TestInvokeExternalStaticMethodArgsWrapped::deserialize(args_buffer, external_module);

  let result = testInvokeExternalStaticMethod(
    args.arg,
  );

  
  StringWrapped::serialize(&result).to_vec()
}

struct TestInvokeExternalStaticMethodArgs {
    pub arg: String,
}
 
impl TestInvokeExternalStaticMethodArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct TestInvokeExternalStaticMethodArgsWrapped {
    pub arg: String,
}

impl TestInvokeExternalStaticMethodArgsWrapped {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }

    pub fn deserialize(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> TestInvokeExternalStaticMethodArgs {
        let args: TestInvokeExternalStaticMethodArgsWrapped = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
       
        TestInvokeExternalStaticMethodArgs::new(
                        
            args.arg,
            
        )
    }  
}
fn invokeTestInvokeExternalInstanceMethodWrapped(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> Vec<u8> {
  let args_buffer = buffer;

  let args = TestInvokeExternalInstanceMethodArgsWrapped::deserialize(args_buffer, external_module);

  let result = testInvokeExternalInstanceMethod(
    args.arg,
  );

  
  StringWrapped::serialize(&result).to_vec()
}

struct TestInvokeExternalInstanceMethodArgs {
    pub arg: String,
}
 
impl TestInvokeExternalInstanceMethodArgs {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct TestInvokeExternalInstanceMethodArgsWrapped {
    pub arg: String,
}

impl TestInvokeExternalInstanceMethodArgsWrapped {
    pub fn new(
        arg: String,
    ) -> Self {
        Self {
            arg,
        }
    }

    pub fn deserialize(buffer: &[u8], external_module: Arc<dyn ExternalModule>) -> TestInvokeExternalInstanceMethodArgs {
        let args: TestInvokeExternalInstanceMethodArgsWrapped = serde_json::from_str(
            str::from_utf8(buffer).expect("Could not convert buffer to string")
        ).expect("JSON was not well-formatted");
       
        TestInvokeExternalInstanceMethodArgs::new(
                        
            args.arg,
            
        )
    }  
}
